File: sdks/rust/src/worker.rs
use crate::context::TaskContext;
use crate::error::{Error, Result};
use crate::types::{ClaimedTask, WorkerOptions};
use parking_lot::RwLock;
use std::collections::HashMap;
// use std::future::Future;
// use std::pin::Pin;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;
use tokio_postgres::Client;
use uuid::Uuid;

// type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;
// type TaskHandler = Arc<dyn Fn(serde_json::Value, TaskContext) -> BoxFuture<'static, Result<serde_json::Value>> + Send + Sync>;

// pub(crate) struct RegisteredTask {
//     pub name: String,
//     pub queue: String,
//     pub handler: TaskHandler,
// }

/// Background worker that polls for and executes tasks
pub struct Worker {
    shutdown_tx: mpsc::Sender<()>,
    worker_handle: JoinHandle<()>,
}

impl Worker {
    pub(crate) async fn new(
        client: Arc<Client>,
        queue_name: String,
        registry: Arc<RwLock<HashMap<String, crate::client::RegisteredTask>>>,
        options: WorkerOptions,
    ) -> Result<Self> {
        let (shutdown_tx, mut shutdown_rx) = mpsc::channel::<()>(1);
        
        let worker_handle = tokio::spawn(async move {
            let mut executing = tokio::task::JoinSet::new();
            let (notify_tx, mut notify_rx) = mpsc::channel::<()>(100);
            
            loop {
                tokio::select! {
                    _ = shutdown_rx.recv() => {
                        tracing::info!("Worker shutting down");
                        break;
                    }
                    
                    _ = notify_rx.recv() => {
                        // Notified of availability, continue to claim tasks
                    }
                    
                    _ = tokio::time::sleep(Duration::from_secs_f64(options.poll_interval_secs)) => {
                        // Poll interval elapsed
                    }
                }
                
                // Check if we have capacity
                if executing.len() >= options.concurrency {
                    if let Some(result) = executing.join_next().await {
                        if let Err(e) = result {
                            tracing::error!("Task execution panicked: {:?}", e);
                        }
                    }
                    continue;
                }
                
                let available_capacity = options.concurrency - executing.len();
                let to_claim = available_capacity.min(options.batch_size);
                
                if to_claim == 0 {
                    continue;
                }
                
                // Claim tasks
                let tasks = match claim_tasks(
                    &client,
                    &queue_name,
                    &options.worker_id,
                    options.claim_timeout,
                    to_claim as i32,
                ).await {
                    Ok(tasks) => tasks,
                    Err(e) => {
                        tracing::error!("Failed to claim tasks: {:?}", e);
                        continue;
                    }
                };
                
                if tasks.is_empty() {
                    continue;
                }
                
                // Spawn task execution
                for task in tasks {
                    let client_clone = Arc::clone(&client);
                    let queue_clone = queue_name.clone();
                    let registry_clone = registry.clone();
                    let claim_timeout = options.claim_timeout;
                    let fatal_on_timeout = options.fatal_on_lease_timeout;
                    let notify_tx_clone = notify_tx.clone();
                    
                    executing.spawn(async move {
                        if let Err(e) = execute_task_with_timeout(
                            client_clone,
                            queue_clone,
                            registry_clone,
                            task,
                            claim_timeout,
                            fatal_on_timeout,
                        ).await {
                            if !e.is_suspended() && !e.is_cancelled() {
                                tracing::error!("Task execution failed: {:?}", e);
                            }
                        }
                        
                        let _ = notify_tx_clone.send(()).await;
                    });
                }
            }
            
            // Wait for all tasks to complete
            while let Some(result) = executing.join_next().await {
                if let Err(e) = result {
                    tracing::error!("Task execution panicked during shutdown: {:?}", e);
                }
            }
        });
        
        Ok(Self {
            shutdown_tx,
            worker_handle,
        })
    }
    
    /// Stop the worker and wait for all tasks to complete
    pub async fn close(self) -> Result<()> {
        let _ = self.shutdown_tx.send(()).await;
        self.worker_handle.await
            .map_err(|e| Error::Other(format!("Worker task panicked: {}", e)))?;
        Ok(())
    }
}

async fn claim_tasks(
    client: &Client,
    queue_name: &str,
    worker_id: &str,
    claim_timeout: i32,
    batch_size: i32,
) -> Result<Vec<ClaimedTask>> {
    let rows = client
        .query(
            "SELECT run_id, task_id, attempt, task_name, params, retry_strategy, 
                    max_attempts, headers, wake_event, event_payload
             FROM absurd.claim_task($1, $2, $3, $4)",
            &[&queue_name, &worker_id, &claim_timeout, &batch_size],
        )
        .await?;
    
    let tasks = rows
        .iter()
        .map(|row| {
            let headers_json: Option<serde_json::Value> = row.get(7);
            let headers: Option<HashMap<String, serde_json::Value>> = headers_json
                .and_then(|v| serde_json::from_value(v).ok());

            ClaimedTask {
                run_id: row.get(0),
                task_id: row.get(1),
                attempt: row.get(2),
                task_name: row.get(3),
                params: row.get(4),
                retry_strategy: row.get(5),
                max_attempts: row.get(6),
                headers,
                wake_event: row.get(8),
                event_payload: row.get(9),
            }
        })
        .collect();
    
    Ok(tasks)
}

async fn execute_task_with_timeout(
    client: Arc<Client>,
    queue_name: String,
    registry: Arc<RwLock<HashMap<String, crate::client::RegisteredTask>>>,
    task: ClaimedTask,
    claim_timeout: i32,
    fatal_on_timeout: bool,
) -> Result<()> {
    let timeout_duration = Duration::from_secs((claim_timeout * 2) as u64);
    
    let result = if fatal_on_timeout {
        tokio::time::timeout(
            timeout_duration,
            execute_task(client, queue_name, registry, task, claim_timeout),
        )
        .await
        .unwrap_or_else(|_| {
            tracing::error!(
                "Task exceeded claim timeout by 2x ({}s), terminating process",
                claim_timeout
            );
            std::process::exit(1);
        })
    } else {
        tokio::time::timeout(
            timeout_duration,
            execute_task(client, queue_name, registry, task, claim_timeout),
        )
        .await
        .unwrap_or_else(|_| {
            tracing::warn!("Task exceeded claim timeout by 2x ({}s)", claim_timeout);
            Err(Error::Other("Task timeout".to_string()))
        })
    };
    
    result
}

async fn execute_task(
    client: Arc<Client>,
    queue_name: String,
    registry: Arc<RwLock<HashMap<String, crate::client::RegisteredTask>>>,
    task: ClaimedTask,
    claim_timeout: i32,
) -> Result<()> {
    let (handler, task_queue) = {
        let registry_lock = registry.read();
        let registered = registry_lock
            .get(&task.task_name)
            .ok_or_else(|| Error::TaskNotRegistered(task.task_name.clone()))?;
        
        (registered.handler.clone(), registered.queue.clone())
    };
    
    // Create context
    let ctx = TaskContext::new(
        client.clone(),
        task_queue,
        task.clone(),
        claim_timeout,
    ).await?;
    
    // Execute handler
    match handler(task.params.clone(), ctx).await {
        Ok(result) => {
            complete_run(&client, &queue_name, &task.run_id, result).await?;
        }
        Err(Error::Suspended) | Err(Error::Cancelled) => {}
        Err(e) => {
            fail_run(&client, &queue_name, &task.run_id, &e).await?;
            return Err(e);
        }
    }
    
    Ok(())
}

async fn complete_run(
    client: &Client,
    queue_name: &str,
    run_id: &Uuid,
    result: serde_json::Value,
) -> Result<()> {
    client
        .execute(
            "SELECT absurd.complete_run($1, $2, $3)",
            &[&queue_name, run_id, &result],
        )
        .await?;
    Ok(())
}

async fn fail_run(
    client: &Client,
    queue_name: &str,
    run_id: &Uuid,
    error: &Error,
) -> Result<()> {
    let error_json = serde_json::json!({
        "name": error.to_string(),
        "message": error.to_string(),
    });
    
    client
        .execute(
            "SELECT absurd.fail_run($1, $2, $3, $4)",
            &[&queue_name, run_id, &error_json, &None::<String>],
        )
        .await?;
    Ok(())
}

File: sdks/go/worker.go
package absurd

import (
	"context"
	"errors"
	"fmt"
	"os"
	"runtime/debug"
	"sync"
	"time"
)

type workerConfig struct {
	batch               WorkBatchOptions
	concurrency         int
	pollInterval        time.Duration
	fatalOnLeaseTimeout bool
	onError             func(error)
}

func normalizeWorkBatchOptions(options WorkBatchOptions) WorkBatchOptions {
	options.WorkerID = orString(options.WorkerID, defaultBatchWorkerID)
	options.ClaimTimeout = durationOr(options.ClaimTimeout, defaultClaimTimeout)
	options.BatchSize = positiveOr(options.BatchSize, 1)
	return options
}

func (c *Client) normalizeWorkerOptions(options WorkerOptions) workerConfig {
	workerID := options.WorkerID
	if workerID == "" {
		hostname, _ := os.Hostname()
		workerID = fmt.Sprintf("%s:%d", hostname, os.Getpid())
	}
	concurrency := positiveOr(options.Concurrency, 1)
	onError := options.OnError
	if onError == nil {
		onError = func(err error) {
			c.logger.Printf("worker error: %v", err)
		}
	}
	fatalOnLeaseTimeout := true
	if options.FatalOnLeaseTimeout != nil {
		fatalOnLeaseTimeout = *options.FatalOnLeaseTimeout
	}
	return workerConfig{
		batch: normalizeWorkBatchOptions(WorkBatchOptions{
			WorkerID:     workerID,
			ClaimTimeout: options.ClaimTimeout,
			BatchSize:    positiveOr(options.BatchSize, concurrency),
		}),
		concurrency:         concurrency,
		pollInterval:        durationOr(options.PollInterval, defaultWorkerPollInterval),
		fatalOnLeaseTimeout: fatalOnLeaseTimeout,
		onError:             onError,
	}
}

func sleepContext(ctx context.Context, d time.Duration) error {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func waitCapacityOrPoll(ctx context.Context, capacityReleased <-chan struct{}, d time.Duration) error {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-capacityReleased:
		return nil
	case <-timer.C:
		return nil
	}
}

func (c *Client) claimTasks(ctx context.Context, options WorkBatchOptions) ([]claimedTask, error) {
	opts := normalizeWorkBatchOptions(options)

	rows, err := c.db.QueryContext(ctx, `SELECT run_id, task_id, attempt, task_name, params, headers, wake_event, event_payload
	       FROM absurd.claim_task($1, $2, $3, $4)`, c.queueName, opts.WorkerID, durationSeconds(opts.ClaimTimeout), opts.BatchSize)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var tasks []claimedTask
	for rows.Next() {
		var task claimedTask
		var headers []byte
		var wakeEvent *string
		var eventPayload []byte
		if err := rows.Scan(
			&task.RunID,
			&task.TaskID,
			&task.Attempt,
			&task.TaskName,
			&task.ParamsRaw,
			&headers,
			&wakeEvent,
			&eventPayload,
		); err != nil {
			return nil, err
		}
		task.HeadersRaw = cloneRawJSON(headers)
		if wakeEvent != nil {
			task.WakeEvent = *wakeEvent
		}
		task.EventRaw = cloneRawJSON(eventPayload)
		tasks = append(tasks, task)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return tasks, nil
}

func (c *Client) WorkBatch(ctx context.Context, options ...WorkBatchOptions) error {
	opts := normalizeWorkBatchOptions(first(options))
	tasks, err := c.claimTasks(ctx, opts)
	if err != nil {
		return err
	}
	for _, task := range tasks {
		if err := c.executeTask(context.WithoutCancel(ctx), task, opts.ClaimTimeout, false); err != nil {
			return err
		}
	}
	return nil
}

func (c *Client) RunWorker(ctx context.Context, options ...WorkerOptions) error {
	cfg := c.normalizeWorkerOptions(first(options))
	sem := make(chan struct{}, cfg.concurrency)
	capacityReleased := make(chan struct{}, 1)
	var wg sync.WaitGroup
	defer wg.Wait()

	for {
		if err := ctx.Err(); err != nil {
			return err
		}

		available := cfg.concurrency - len(sem)
		if available == 0 {
			if err := waitCapacityOrPoll(ctx, capacityReleased, cfg.pollInterval); err != nil {
				return err
			}
			continue
		}

		batch := cfg.batch
		if batch.BatchSize > available {
			batch.BatchSize = available
		}
		tasks, err := c.claimTasks(ctx, batch)
		if err != nil {
			if ctx.Err() != nil {
				return ctx.Err()
			}
			cfg.onError(err)
			if err := sleepContext(ctx, cfg.pollInterval); err != nil {
				return err
			}
			continue
		}
		if len(tasks) == 0 {
			if err := sleepContext(ctx, cfg.pollInterval); err != nil {
				return err
			}
			continue
		}

		for _, task := range tasks {
			sem <- struct{}{}
			wg.Add(1)
			go func(task claimedTask, claimTimeout time.Duration) {
				defer wg.Done()
				defer func() {
					<-sem
					select {
					case capacityReleased <- struct{}{}:
					default:
					}
				}()
				if err := c.executeTask(context.WithoutCancel(ctx), task, claimTimeout, cfg.fatalOnLeaseTimeout); err != nil {
					cfg.onError(err)
				}
			}(task, batch.ClaimTimeout)
		}
	}
}

func panicDetails(v any) (string, string) {
	if err, ok := v.(error); ok {
		return fmt.Sprintf("%T", err), fmt.Sprintf("panic: %v", err)
	}
	return fmt.Sprintf("%T", v), fmt.Sprintf("panic: %v", v)
}

type leaseWatchdog struct {
	mu                  sync.Mutex
	warn                *time.Timer
	fatal               *time.Timer
	stopped             bool
	epoch               uint64
	logger              Logger
	taskLabel           string
	fatalOnLeaseTimeout bool
}

func newLeaseWatchdog(logger Logger, taskName string, taskID string, fatalOnLeaseTimeout bool) *leaseWatchdog {
	return &leaseWatchdog{
		logger:              logger,
		taskLabel:           fmt.Sprintf("%s (%s)", taskName, taskID),
		fatalOnLeaseTimeout: fatalOnLeaseTimeout,
	}
}

func (w *leaseWatchdog) schedule(lease time.Duration) {
	if lease <= 0 {
		return
	}
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.stopped {
		return
	}
	w.epoch++
	epoch := w.epoch
	if w.warn != nil {
		w.warn.Stop()
	}
	if w.fatal != nil {
		w.fatal.Stop()
	}
	w.warn = time.AfterFunc(lease, func() {
		w.mu.Lock()
		if w.stopped || w.epoch != epoch {
			w.mu.Unlock()
			return
		}
		w.mu.Unlock()
		w.logger.Printf("[absurd] task %s exceeded claim timeout of %s", w.taskLabel, lease)
	})
	if !w.fatalOnLeaseTimeout {
		return
	}
	w.fatal = time.AfterFunc(2*lease, func() {
		w.mu.Lock()
		if w.stopped || w.epoch != epoch {
			w.mu.Unlock()
			return
		}
		w.mu.Unlock()
		w.logger.Printf("[absurd] task %s exceeded claim timeout of %s by more than 100%%; terminating process", w.taskLabel, lease)
		os.Exit(1)
	})
}

func (w *leaseWatchdog) stop() {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.stopped = true
	w.epoch++
	if w.warn != nil {
		w.warn.Stop()
		w.warn = nil
	}
	if w.fatal != nil {
		w.fatal.Stop()
		w.fatal = nil
	}
}

func (c *Client) executeTask(ctx context.Context, task claimedTask, claimTimeout time.Duration, fatalOnLeaseTimeout bool) (err error) {
	completionCtx := context.WithoutCancel(ctx)
	watchdog := newLeaseWatchdog(c.logger, task.TaskName, task.TaskID, fatalOnLeaseTimeout)
	defer watchdog.stop()
	watchdog.schedule(claimTimeout)

	defer func() {
		if v := recover(); v != nil {
			name, message := panicDetails(v)
			c.logger.Printf("[absurd] task execution panicked: %s", message)
			err = failTaskRunWithTraceback(completionCtx, c.db, c.queueName, task.RunID, name, message, debug.Stack())
		}
	}()

	registration, ok := c.getRegistration(task.TaskName)
	if !ok {
		err := fmt.Errorf("unknown task %q", task.TaskName)
		c.logger.Printf("[absurd] %v", err)
		return failTaskRun(completionCtx, c.db, c.queueName, task.RunID, err)
	}
	if registration.queueName != c.queueName {
		err := fmt.Errorf("misconfigured task %q (queue mismatch)", task.TaskName)
		c.logger.Printf("[absurd] %v", err)
		return failTaskRun(completionCtx, c.db, c.queueName, task.RunID, err)
	}
	taskCtx, err := newTaskContext(ctx, c, registration.queueName, task, claimTimeout, func(d time.Duration) {
		watchdog.schedule(d)
	})
	if err != nil {
		if errors.Is(err, errInvalidTaskHeaders) {
			c.logger.Printf("[absurd] %v", err)
			return failTaskRun(completionCtx, c.db, c.queueName, task.RunID, err)
		}
		return err
	}
	execute := func() (any, error) {
		return registration.handler(taskCtx, task.ParamsRaw)
	}
	var result any
	if c.hooks.WrapTaskExecution != nil {
		result, err = c.hooks.WrapTaskExecution(taskCtx, execute)
	} else {
		result, err = execute()
	}
	if err != nil {
		if errors.Is(err, errSuspend) || errors.Is(err, errCancelled) || errors.Is(err, errFailedRun) {
			return nil
		}
		c.logger.Printf("[absurd] task execution failed: %v", err)
		return failTaskRun(completionCtx, c.db, c.queueName, task.RunID, err)
	}
	return completeTaskRun(completionCtx, c.db, c.queueName, task.RunID, result)
}
